### PR TITLE
ddl: support report error when replicate incompatible ddl

### LINKED ISSUE #9850: dm: support report error when replicate incompatible ddl
### Is your feature request related to a problem?

Some customers have experienced issues with upstream tools like pt-osc, which truncate the range of values during DDL changes like modifying Varchar(64) to Varchar(32), resulting in data loss. While these DDL changes might only produce warnings, there is a concern that they could execute successfully in TiDB under certain SQL modes and potentially cause downstream failures. It is advisable for DM to intercept and handle these DDL changes proactively. This approach also prevents prolonged scanning of large tables where eventual errors might occur due to data length mismatches.

### Describe the feature you'd like

```
filters:
  rule-1:
    schema-pattern: "*"   # The name of the database for which this binlog event filter should be applied. Use "*" to indicate all databases, as needed by the user.
    table-pattern: "*"    # The name of the table for which this binlog event filter should be applied. Use "*" to indicate all tables, as needed by the user.
    events: ["incompatible_ddl_changes"]  # The binlog events to be parsed. Set to "incompatible_ddl_changes" to represent DDL changes that cause incompatibility and data inconsistency.
    action: error  # The action to be taken when this DDL is encountered. "error" indicates an error should be raised, and synchronization should stop. Other possible values are "ignore" and "do". The default value is "do".

  rule-2:
    schema-pattern: "test1"   # The name of the database for which this binlog event filter should be applied, set to "test1".
    table-pattern: "t*"    # The name of the table for which this binlog event filter should be applied, for tables starting with "t".
    events: ["Modify Column", "Value Range Decrease"]  # Represents the event types to be parsed and blocked. Here, it indicates modifying column types and reducing the value range.
    action: error  # The action to be taken when this DDL is encountered. "error" indicates an error should be raised, and synchronization should stop. Other possible values are "ignore" and "do". The default value is "do".

```

### Describe alternatives you've considered

_No response_

### Teachability, Documentation, Adoption, Migration Strategy

_No response_

### PR DESCRIPTION
**Pull request by GMHDBJD**
### What problem does this PR solve?

Issue Number: close #9850 

### What is changed and how it works?
- judge incompatible ddl changes for alter table

#### Tests 

 - Integration test

### Release note 

```release-note
Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

If you don't think this PR needs a release note then fill it with `None`.
```

### CODE REVIEW

**Review by lichunzhu — APPROVED** [review]
LGTM

File: dm/syncer/ddl.go [hunk 1]
```diff
@@ -955,14 +965,228 @@ func parseOneStmt(qec *queryEventContext) (stmt ast.StmtNode, err error) {
...
+			return bf.AlterTable
+		case mysql.TypeString:
+			// Due to the behavior of padding \x00 at binary type, always change column data when binary length changed
+			if types.IsBinaryStr(&oldCol.FieldType) {
+				if newCol.GetFlen() != oldCol.GetFlen() {
+					return bf.PrecisionDecrease
+				}
+			}
+			return bf.AlterTable
+		}
```
Comment (okJiang) [inline dm/syncer/ddl.go:1011]: These are all cases of inequality, not cases of reduced precision?

File: dm/tests/incompatible_ddl_changes/run.sh [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+	dmctl_start_task_standalone $cur/conf/$1
+
+	run_sql_file $cur/data/db1.increment.sql $MYSQL_HOST1 $MYSQL_PORT1 $MYSQL_PASSWORD1
+
+	run_dm_ctl_with_retry $WORK_DIR "127.0.0.1:$MASTER_PORT" \
+		"query-status test" \
+		"\"result\": true" 2 \
+		"\"synced\": true" 1
+
+	# value range decrease
```
Comment (okJiang) [inline dm/tests/incompatible_ddl_changes/run.sh:31]: can you add the original column type in the comment? It is convenient.

**Review by okJiang — COMMENTED** [review]
rest lgtm

File: dm/tests/incompatible_ddl_changes/run.sh (hunk 2 above)
Comment (GMHDBJD) [reply dm/tests/incompatible_ddl_changes/run.sh:31]: you can see from c_mediumint smallint(6), the column name descripe that it's a mediumint column, also can see the revert ddl at L39, which revert it to mediumint(7)

**Review by okJiang — APPROVED** [review]

**Pull request merged**